netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java

```diff
@@ -501,4 +505,128 @@ public void transportTracer_windowUpdate_local() throws Exception {
+    assertEquals(3, handler.flowControlPing().getPingCount());
+  }
+
+  @Test
+  public void testPingBackoff() throws Exception {
+    AbstractNettyHandler handler = setupPingTest();
+    long pingData = handler.flowControlPing().payload();
+    ByteBuf data40KbBuf = initXkbBuffer(40);
+
+    handler.flowControlPing().setDataSizeAndSincePing(200000);
+
+    for (int i = 0; i <= 10; i++) {
+      int beforeCount = handler.flowControlPing().getPingCount();
+      // should resize on 0
+      readXCopies(6, data40KbBuf); // initiate ping on i= {0, 1, 3, 6, 10}
+      int afterCount = handler.flowControlPing().getPingCount();
+      fakeClock().forwardNanos(200);
+      if (afterCount > beforeCount) {
+        readPingAck(pingData); // should increase backoff multiplier
+      }
+    }
+    assertEquals(5, handler.flowControlPing().getPingCount());
+  }
+
+  @Test
+  public void bdpPingWindowResizing() throws Exception {
+    this.flowControlWindow = 1024 * 100;
+    manualSetUp();
+    makeStream();
+
+    AbstractNettyHandler handler = (AbstractNettyHandler) handler();
+    handler.setAutoTuneFlowControl(true);
+    Http2LocalFlowController localFlowController = connection().local().flowController();
+    long pingData = handler.flowControlPing().payload();
+    int initialWindowSize = localFlowController.initialWindowSize();
+    ByteBuf data1Kb = initXkbBuffer(1);
+    ByteBuf data40Kb = initXkbBuffer(40);
+
+    readXCopies(1, data1Kb); // initiate ping
+    fakeClock().forwardNanos(2);
+    readPingAck(pingData); // should not resize window because of small target window
+    assertEquals(initialWindowSize, localFlowController.initialWindowSize());
+
+    readXCopies(2, data40Kb); // initiate ping
+    fakeClock().forwardNanos(200);
+    readPingAck(pingData); // should resize window
+    int windowSizeA = localFlowController.initialWindowSize();
+    Assert.assertNotEquals(initialWindowSize, windowSizeA);
+
+    readXCopies(4, data40Kb); // initiate ping
+    fakeClock().forwardNanos(400);
+    readPingAck(pingData); // should not resize window as bandwidth didn't increase
+    Assert.assertEquals(windowSizeA, localFlowController.initialWindowSize());
+
+    readXCopies(4, data40Kb); // initiate ping
+    fakeClock().forwardNanos(200);
+    readPingAck(pingData); // should resize window
+    int windowSizeB = localFlowController.initialWindowSize();
+    Assert.assertNotEquals(windowSizeA, windowSizeB);
+  }
+
+
+  private void readPingAck(long pingData) throws Exception {
+    channelRead(pingFrame(true, pingData));
+  }
+
+  private void readXCopies(int copies, ByteBuf data) throws Exception {
+    for (int i = 0; i < copies; i++) {
+      channelRead(dataFrame(STREAM_ID, false, data.copy())); // buffer it
+      stream().request(1); // consume it
+    }
+  }
+
+  private ByteBuf initXkbBuffer(int multiple) {
+    ByteBuf data = ctx().alloc().buffer(5 + 1024 * multiple);
+    // add header
+    data.writeByte(0);
```

Comment:
Seems this should use `dataFrame()`

Reply:
This is to create a copyable byte buffer for use with dataFrame.

Reply:
Ah, I see. Did you consider using `grpcDataFrame()` instead of `dataFrame()`?

Reply:
grpcDataFrame takes a byte[] which I could return instead of the ByteBuf, but would still need to be built the same way.  grpcDataFrame probably would be slightly more efficient which makes it a good practice.  I'll update it.

Reply:
With `byte[]` I figured you'd use `new byte[1024*x]` instead of `initXkbBuffer()`, what whatever.